### ISSUE TITLE
no formal definition of "thenable"

### ISSUE THREAD

**Issue by ikravets** [tc39/ecma262#2868]
It seems like the spec uses the word "thenable" several times, but it's never formally defined. See for example
- [27.1.1.4 The AsyncIterator Interface](https://tc39.es/ecma262/#sec-asynciterator-interface)
- [27.2.2.2 NewPromiseResolveThenableJob ( promiseToResolve, thenable, then )](https://tc39.es/ecma262/#sec-newpromiseresolvethenablejob)

I think it's better to provide an exact definition for  "thenable".

**Comment by bakkot** [comment]
Ehhh, defining it precisely is tricky - what does it mean to "have a method", precisely, in a world with proxies?

**Comment by raulsebastianmihaila** [comment]
> what does it mean to "have a method"
The spec defines a method as a function that is the value of a property of an object (with different words). Where is the challenge?

**Comment by michaelficarra** [comment]
@raulsebastianmihaila well then that term couldn't be used when defining "thenable", could it?

**Comment by raulsebastianmihaila** [comment]
Of course it can be used, what's the challenge?

**Comment by michaelficarra** [comment]
@raulsebastianmihaila again, it is proxies

**Comment by bathos** [comment]
@michaelficarra and even just accessors, or ordinary objects inheriting from proxies or objects with `then` accessors, etc. For that matter, “thenable module exotic objects” or even “thenable sloppy mode arguments exotic objects”!).

The spec does have some pseudo-definitions with these same problems currently, i.e. attempts to provide “shape” definitions for things that are, at runtime, actually categories of _programs,_ not categories of values. I think it probably shouldn’t (or that they should be moved into a non-normative glossary), but I don’t think this would be without precedent.

In runtime terms, the only sound definition of a “thenable value” is _an object,_ nothing more. The runtime definition of interest occurs in the form of algorithmic steps from which the authorial intention-non-runtime-type-notion of “thenable” being something that emerges from it. Unbranded contracts describe categories of _programs,_ not a categories of values. To soundly describe the author-intent notion of a “thenable value” in runtime terms is [definitionally impossible](https://en.wikipedia.org/wiki/Halting_problem).

I don’t think that means there’s nowhere a definition of this sort could live or that adding it lacks value — just think it’s important to be clear that there’s no way a definition of this sort can describe the real set of interest which is like “application states wherein a given value would, if immediately asked to, sing a beautiful rendition _then song._” For this reason I think adding any qualifiers to a definition like the one given in this PR would make it _worse,_ not better: the starting point is describing an author intuition plainly, but the set of qualifiers one would need to “make it sound” is infinitely complex, so better to have none at all than a few rando ones probably?

**Comment by raulsebastianmihaila** [comment]
> again, it is proxies

You're very cryptic. What's the challenge with proxies?

The definition from #2869 is pretty close. A 'thenable' is an object that has a property (own or inherited, this extra wording doesn't need to be part of the definition) whose key is 'then' and whose value is a function. 'Has a property' here needs to be understood in terms of observing that the object has that property (and there are a few different ways to do that). It seems you're suggesting there's an issue with this definition because of the dynamic aspect of the JS language. I don't see any issue with that. JS is a dynamic and stateful language. An object (depending on its state) can be at one point a thenable and afterwards it can stop being a thenable. There are also some invariants and tools in the language to make things more predictable, such that a thenable object can continue to be forever a thenable. For instance, if the `then` property is an own data property, you can freeze the object. If it's an own data property of the obect's prototype, you can prevent extensions on the object and freeze the prototype object. There is of course also the case of the getter, in which case you can not know in advance, without knowing the algorithm in that getter, what the value of the property will be and that's fine. It is enough to observe at any point whether the object has that property and its value. Similar to the definition of the word 'method', as explained in #2846, things need to be understood in terms of relationships and not only in terms of what they consist of intrinsically. And to be maximally clear, the value can be observed using the `[[Get]]` operation (for instance by evaluating a property access), or using a `[[GetOwnProperty]]` operation and looking at either the `value` attribute or at the value of the 'get' attribute, if it's `undefined`, or at the result of calling the value of the `get` attribute, if it's a function.

**Comment by raulsebastianmihaila** [comment]
Actually, reading the definition from #2869 more carefully, I see it's good enough. Another way to express it would be: a thenable is an object that has a method whose key is 'then'.

**Comment by devsnek** [comment]
why do proxies complicate this? if they're well behaved then they either conform to the definition or they don't, like a normal object. if they aren't well behaved then it doesn't matter either way?

**Comment by bathos** [comment]
> It seems [@michaelficarra is] suggesting there's an issue with this definition because of the dynamic aspect of the JS language. I don't see any issue with that. JS is a dynamic and stateful language.

That’s the zone where the `meta-issue` lives, yep.

> An object (depending on its state) can be at one point a thenable and afterwards it can stop being a thenable.

From the POV of algorithmic steps effecting the illusion of that category existing, the category does not exist, so this isn’t true. Oops, getting sucked in, sorry —

The meta layering problem thingie can’t be reasoned away, but it already exists in the spec with or without one more example of it. So long as there continue to be no references to the afflicted definitions in algorithmic steps or other specs, “technically unsound” gets pronounced with a hard technically. They provide value by documenting the higher-level semantic intentions of contractual algorithms.

> things need to be understood in terms of relationships

Definitions for ~higher-level semantic intentions of contractual algorithms~ 🤔 realized relationships should definitely prefer to reference one another. 

> It is enough to observe at any point whether the object has that property and its value. 

ahhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh stop temptttting me nooooo this isn’t actually relevant but <sup>evaluating the contract (a program) can call into user code. it cannot be skipped in these examples and it can always impact application state and it can always throw even if you previously established an invariant about the only possible _normal_ completion value one of its OIMs can return</sup>

**Comment by raulsebastianmihaila** [comment]
bathos's comment looks like word salad (and potentially trolling) to me.

**Comment by bathos** [comment]
@raulsebastianmihaila I can assure you there’s no trolling. I can offer no assurance about its quality, of course, and am sorry you found it lacking.

**Comment by zloirock** [comment]
I absolutely agree with @raulsebastianmihaila. Proxies do not change anything. If it's an object and on getting a `.then` property from an object, you get a callable object - it's a thenable - and here makes no sense is this a real property or not, and will this property be available after getting / calling or will it be removed after that - anyway, "thenable" is something about a duck typing.

> If it walks like a duck and it quacks like a duck, then it must be a duck

**Comment by bathos** [comment]
If any folks have ended up tackling ghosts and/or windmills because my attempt at describing an existing formalization problem only indirectly related to this PR was unsuccessful, apologies. I’m think both @bakkot and @michaelficarra were speaking about that problem, but am not sure of that. If so, maybe a distinct issue (with a description written by anyone but me lol) would prevent further confusion?

**Comment by michaelficarra** [comment]
> If it's an object and on getting a `.then` property from this object, you get a callable object - it's a thenable

Right, so is that the wording you think we should use for the definition?

**Comment by zloirock** [comment]
> Right, so is that the wording you think we should use for the definition?

The definition from #2869 LGTM. However, if it confuses you - it can be changed to something like what you quoted.
